Python 3 compatibility

Hi,

while `pip install fastcsv` works seemlessly on my Python 2.7x installation on Ubuntu 14.04 (with `gcc` and `python-dev` installed), I cannot manage to install it on Python 3.5.1. The installation process almost immediately fails with the following output:

```
Collecting fastcsv
  Using cached fastcsv-0.1.2.tar.gz
Building wheels for collected packages: fastcsv
  Running setup.py bdist_wheel for fastcsv ... error
  Complete output from command /usr/local/bin/python3.5 -u -c "import setuptools, tokenize;__file__='/tmp/pip-build-ncqqpon6/fastcsv/setup.py';exec(compile(getattr(tokenize, 'open', open)(__f
ile__).read().replace('\r\n', '\n'), __file__, 'exec'))" bdist_wheel -d /tmp/tmp82902pxipip-wheel- --python-tag cp35:
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.linux-x86_64-3.5
  copying fastcsv.py -> build/lib.linux-x86_64-3.5
  running build_ext
  building '_fastcsv' extension
  creating build/temp.linux-x86_64-3.5
  gcc -pthread -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/usr/local/include/python3.5m -c _fastcsv.c -o build/temp.linux-x
86_64-3.5/_fastcsv.o
  _fastcsv.c: In function 'init_fastcsv':
  _fastcsv.c:37:38: warning: 'return' with no value, in function returning non-void [-Wreturn-type]
     if (PyType_Ready(&ReaderType) < 0) return;
                                        ^
  _fastcsv.c:38:38: warning: 'return' with no value, in function returning non-void [-Wreturn-type]
     if (PyType_Ready(&WriterType) < 0) return;
                                        ^
  _fastcsv.c:42:5: warning: implicit declaration of function 'Py_InitModule' [-Wimplicit-function-declaration]
       m = Py_InitModule("_fastcsv", _fastcsv_methods);
       ^
  _fastcsv.c:42:7: warning: assignment makes pointer from integer without a cast [enabled by default]
       m = Py_InitModule("_fastcsv", _fastcsv_methods);
         ^
  gcc -pthread -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/usr/local/include/python3.5m -c _fastcsv_reader.c -o build/temp.
linux-x86_64-3.5/_fastcsv_reader.o
  _fastcsv_reader.c: In function 'ParseNewlineMode':
  _fastcsv_reader.c:57:5: warning: implicit declaration of function 'PyString_Check' [-Wimplicit-function-declaration]
       if (PyString_Check(newline)) {
       ^
  _fastcsv_reader.c: In function 'Reader_init':
  _fastcsv_reader.c:112:3: warning: implicit declaration of function 'PyString_FromString' [-Wimplicit-function-declaration]
     self->read_string = PyString_FromString("read");
     ^
  _fastcsv_reader.c:112:21: warning: assignment makes pointer from integer without a cast [enabled by default]
     self->read_string = PyString_FromString("read");
                       ^
  _fastcsv_reader.c:114:3: warning: implicit declaration of function 'PyInt_FromLong' [-Wimplicit-function-declaration]
     self->read_arg = PyInt_FromLong(1024);
     ^
  _fastcsv_reader.c:114:18: warning: assignment makes pointer from integer without a cast [enabled by default]
     self->read_arg = PyInt_FromLong(1024);
                    ^
  _fastcsv_reader.c: At top level:
  _fastcsv_reader.c:505:3: warning: missing braces around initializer [-Wmissing-braces]
     PyObject_HEAD_INIT(NULL)
     ^
  _fastcsv_reader.c:505:3: warning: (near initialization for 'ReaderType.ob_base.ob_base') [-Wmissing-braces]
  _fastcsv_reader.c:507:3: warning: initialization makes integer from pointer without a cast [enabled by default]
     "_fastcsv.Reader",             /* tp_name */
     ^
  _fastcsv_reader.c:507:3: warning: (near initialization for 'ReaderType.tp_basicsize') [enabled by default]
  _fastcsv_reader.c:510:3: warning: initialization from incompatible pointer type [enabled by default]
     (destructor)Reader_dealloc,    /* tp_dealloc */
     ^
  _fastcsv_reader.c:510:3: warning: (near initialization for 'ReaderType.tp_print') [enabled by default]
  _fastcsv_reader.c:525:24: error: 'Py_TPFLAGS_HAVE_ITER' undeclared here (not in a function)
     Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_ITER, /* tp_flags */
                          ^
  _fastcsv_reader.c:526:3: warning: initialization from incompatible pointer type [enabled by default]
     "FastCSV Reader Object",       /* tp_doc */
     ^
  _fastcsv_reader.c:526:3: warning: (near initialization for 'ReaderType.tp_traverse') [enabled by default]
  _fastcsv_reader.c:532:3: warning: initialization from incompatible pointer type [enabled by default]
     (iternextfunc)Reader_iternext, /* tp_iternext */
     ^
  _fastcsv_reader.c:532:3: warning: (near initialization for 'ReaderType.tp_methods') [enabled by default]
  _fastcsv_reader.c:533:3: warning: initialization from incompatible pointer type [enabled by default]
     Reader_methods,                /* tp_methods */
     ^
  _fastcsv_reader.c:533:3: warning: (near initialization for 'ReaderType.tp_members') [enabled by default]
  _fastcsv_reader.c:541:3: warning: initialization from incompatible pointer type [enabled by default]
     (initproc)Reader_init,         /* tp_init */
     ^
  _fastcsv_reader.c:541:3: warning: (near initialization for 'ReaderType.tp_alloc') [enabled by default]
  _fastcsv_reader.c:543:3: warning: initialization from incompatible pointer type [enabled by default]
     PyType_GenericNew,             /* tp_new */
     ^
  _fastcsv_reader.c:543:3: warning: (near initialization for 'ReaderType.tp_free') [enabled by default]
  error: command 'gcc' failed with exit status 1

  ----------------------------------------
  Failed building wheel for fastcsv
  Running setup.py clean for fastcsv
Failed to build fastcsv
Installing collected packages: fastcsv
  Running setup.py install for fastcsv ... error
    Complete output from command /usr/local/bin/python3.5 -u -c "import setuptools, tokenize;__file__='/tmp/pip-build-ncqqpon6/fastcsv/setup.py';exec(compile(getattr(tokenize, 'open', open)(_
_file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-vqqcl5h3-record/install-record.txt --single-version-externally-managed --compile:
    running install
    running build
    running build_py
    creating build
    creating build/lib.linux-x86_64-3.5
    copying fastcsv.py -> build/lib.linux-x86_64-3.5
    running build_ext
    building '_fastcsv' extension
    creating build/temp.linux-x86_64-3.5
    gcc -pthread -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/usr/local/include/python3.5m -c _fastcsv.c -o build/temp.linux
-x86_64-3.5/_fastcsv.o
    _fastcsv.c: In function 'init_fastcsv':
    _fastcsv.c:37:38: warning: 'return' with no value, in function returning non-void [-Wreturn-type]
       if (PyType_Ready(&ReaderType) < 0) return;
                                          ^
    _fastcsv.c:38:38: warning: 'return' with no value, in function returning non-void [-Wreturn-type]
       if (PyType_Ready(&WriterType) < 0) return;
                                          ^
    _fastcsv.c:42:5: warning: implicit declaration of function 'Py_InitModule' [-Wimplicit-function-declaration]
         m = Py_InitModule("_fastcsv", _fastcsv_methods);
         ^
    _fastcsv.c:42:7: warning: assignment makes pointer from integer without a cast [enabled by default]
         m = Py_InitModule("_fastcsv", _fastcsv_methods);
           ^
    gcc -pthread -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/usr/local/include/python3.5m -c _fastcsv_reader.c -o build/tem
p.linux-x86_64-3.5/_fastcsv_reader.o
    _fastcsv_reader.c: In function 'ParseNewlineMode':
    _fastcsv_reader.c:57:5: warning: implicit declaration of function 'PyString_Check' [-Wimplicit-function-declaration]
         if (PyString_Check(newline)) {
         ^
    _fastcsv_reader.c: In function 'Reader_init':
    _fastcsv_reader.c:112:3: warning: implicit declaration of function 'PyString_FromString' [-Wimplicit-function-declaration]
       self->read_string = PyString_FromString("read");
       ^
    _fastcsv_reader.c:112:21: warning: assignment makes pointer from integer without a cast [enabled by default]
       self->read_string = PyString_FromString("read");
                         ^
    _fastcsv_reader.c:114:3: warning: implicit declaration of function 'PyInt_FromLong' [-Wimplicit-function-declaration]
       self->read_arg = PyInt_FromLong(1024);
       ^
    _fastcsv_reader.c:114:18: warning: assignment makes pointer from integer without a cast [enabled by default]
       self->read_arg = PyInt_FromLong(1024);
                      ^
    _fastcsv_reader.c: At top level:
    _fastcsv_reader.c:505:3: warning: missing braces around initializer [-Wmissing-braces]
       PyObject_HEAD_INIT(NULL)
       ^
    _fastcsv_reader.c:505:3: warning: (near initialization for 'ReaderType.ob_base.ob_base') [-Wmissing-braces]
    _fastcsv_reader.c:507:3: warning: initialization makes integer from pointer without a cast [enabled by default]
       "_fastcsv.Reader",             /* tp_name */
       ^
    _fastcsv_reader.c:507:3: warning: (near initialization for 'ReaderType.tp_basicsize') [enabled by default]
    _fastcsv_reader.c:510:3: warning: initialization from incompatible pointer type [enabled by default]
       (destructor)Reader_dealloc,    /* tp_dealloc */
       ^
    _fastcsv_reader.c:510:3: warning: (near initialization for 'ReaderType.tp_print') [enabled by default]
    _fastcsv_reader.c:525:24: error: 'Py_TPFLAGS_HAVE_ITER' undeclared here (not in a function)
       Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_ITER, /* tp_flags */
                            ^
    _fastcsv_reader.c:526:3: warning: initialization from incompatible pointer type [enabled by default]
       "FastCSV Reader Object",       /* tp_doc */
       ^
    _fastcsv_reader.c:526:3: warning: (near initialization for 'ReaderType.tp_traverse') [enabled by default]
    _fastcsv_reader.c:532:3: warning: initialization from incompatible pointer type [enabled by default]
       (iternextfunc)Reader_iternext, /* tp_iternext */
       ^
    _fastcsv_reader.c:532:3: warning: (near initialization for 'ReaderType.tp_methods') [enabled by default]
    _fastcsv_reader.c:533:3: warning: initialization from incompatible pointer type [enabled by default]
       Reader_methods,                /* tp_methods */
       ^
    _fastcsv_reader.c:533:3: warning: (near initialization for 'ReaderType.tp_members') [enabled by default]
    _fastcsv_reader.c:541:3: warning: initialization from incompatible pointer type [enabled by default]
       (initproc)Reader_init,         /* tp_init */
       ^
    _fastcsv_reader.c:541:3: warning: (near initialization for 'ReaderType.tp_alloc') [enabled by default]
    _fastcsv_reader.c:543:3: warning: initialization from incompatible pointer type [enabled by default]
       PyType_GenericNew,             /* tp_new */
       ^
    _fastcsv_reader.c:543:3: warning: (near initialization for 'ReaderType.tp_free') [enabled by default]
    error: command 'gcc' failed with exit status 1

    ----------------------------------------
Command "/usr/local/bin/python3.5 -u -c "import setuptools, tokenize;__file__='/tmp/pip-build-ncqqpon6/fastcsv/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace(
'\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-vqqcl5h3-record/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /tmp/pip-build-ncqqp
on6/fastcsv/
```

Is there any possibility to also install the library on Python 3.x?

--- a/_fastcsv.c
+++ b/_fastcsv.c
@@ -32,6 +32,35 @@ static PyMethodDef _fastcsv_methods[] = {
   {NULL}
 };
 
+#if PY_MAJOR_VERSION >= 3
+static struct PyModuleDef moduledef = {
+  PyModuleDef_HEAD_INIT,
+  "_fastcsv",
+  NULL,
+  0,
+  _fastcsv_methods,
+  NULL,
+  NULL,
+  NULL,
+  NULL,
+};
+
+PyMODINIT_FUNC
+PyInit__fastcsv(void) {
+  if (PyType_Ready(&ReaderType) < 0) return NULL;
+  if (PyType_Ready(&WriterType) < 0) return NULL;
+
+  PyObject *m = PyModule_Create(&moduledef);
+  if (m == NULL) {
+    return NULL;
+  }
+  Py_INCREF(&ReaderType);
+  PyModule_AddObject(m, "Reader", (PyObject *)&ReaderType);
+  Py_INCREF(&WriterType);
+  PyModule_AddObject(m, "Writer", (PyObject *)&WriterType);
+  return m;
+}
+#else
 PyMODINIT_FUNC
 init_fastcsv(void) {
   if (PyType_Ready(&ReaderType) < 0) return;
@@ -47,4 +76,4 @@ init_fastcsv(void) {
     PyModule_AddObject(m, "Writer", (PyObject *)&WriterType);
   }
 }
-
+#endif

--- a/_fastcsv_reader.c
+++ b/_fastcsv_reader.c
@@ -54,6 +54,10 @@ ParseNewlineMode(PyObject *newline, NewlineMode *newline_mode) {
     *newline_mode = UniversalNewline;
   } else {
     unsigned char free_newline = 0;
+#if PY_MAJOR_VERSION >= 3
+    // Do not do bytes to unicode conversion in Python3. We expect the caller to
+    // use unicode strings consistently in Python3.
+#else
     if (PyString_Check(newline)) {
       PyObject *decoded = PyUnicode_FromObject(newline);
       if (!decoded) {
@@ -63,6 +67,7 @@ ParseNewlineMode(PyObject *newline, NewlineMode *newline_mode) {
       free_newline = 1;
       newline = decoded;
     }
+#endif
     if (PyUnicode_Check(newline)) {
       if (PyUnicode_GET_SIZE(newline) == 1) {
         if (PyUnicode_AS_UNICODE(newline)[0] == '\r') {
@@ -109,9 +114,13 @@ Reader_init(Reader *self, PyObject *args, PyObject *kwds) {
   self->contents = PyMem_New(PyObject *, self->content_cap);
   if (!self->contents) goto error;
 
+#if PY_MAJOR_VERSION >= 3
+  self->read_string = PyUnicode_FromString("read");
+#else
   self->read_string = PyString_FromString("read");
+#endif
   if (!self->read_string) goto error;
-  self->read_arg = PyInt_FromLong(1024);
+  self->read_arg = PyLong_FromLong(1024);
   if (!self->read_arg) goto error;
 
   self->entered = 0;
@@ -502,8 +511,7 @@ static PyMethodDef Reader_methods[] = {
 };
 
 PyTypeObject ReaderType = {
-  PyObject_HEAD_INIT(NULL)
-  0,                             /* ob_size */
+  PyVarObject_HEAD_INIT(NULL, 0)
   "_fastcsv.Reader",             /* tp_name */
   sizeof(Reader),                /* tp_basicsize */
   0,                             /* tp_itemsize */
@@ -522,7 +530,13 @@ PyTypeObject ReaderType = {
   0,                             /* tp_getattro */
   0,                             /* tp_setattro */
   0,                             /* tp_as_buffer */
-  Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_ITER, /* tp_flags */
+  Py_TPFLAGS_DEFAULT
+#if PY_MAJOR_VERSION >= 3
+    // Py_TPFLAGS_HAVE_ITER is deprecated in Python3.
+#else
+    | Py_TPFLAGS_HAVE_ITER
+#endif
+    , /* tp_flags */
   "FastCSV Reader Object",       /* tp_doc */
   0,                             /* tp_traverse */
   0,                             /* tp_clear */

--- a/_fastcsv_writer.c
+++ b/_fastcsv_writer.c
@@ -223,7 +223,11 @@ Writer_writecell(Writer *self, PyObject *cell,
     }
     return 1;
   } else {
+#if PY_MAJOR_VERSION >= 3
+    cellstr = PyObject_Str(cell);
+#else
     cellstr = PyObject_Unicode(cell);
+#endif
     if (!cellstr) {
       PyErr_SetString(PyExc_ValueError, "cell value is not unicode");
       goto error_exit;
@@ -380,8 +384,7 @@ static PyMethodDef Writer_methods[] = {
 };
 
 PyTypeObject WriterType = {
-  PyObject_HEAD_INIT(NULL)
-  0,                          /* ob_size */
+  PyVarObject_HEAD_INIT(NULL, 0)
   "_fastcsv.Writer",          /* tp_name */
   sizeof(Writer),             /* tp_basicsize */
   0,                          /* tp_itemsize */
@@ -400,7 +403,13 @@ PyTypeObject WriterType = {
   0,                          /* tp_getattro */
   0,                          /* tp_setattro */
   0,                          /* tp_as_buffer */
-  Py_TPFLAGS_DEFAULT,         /* tp_flags */
+  Py_TPFLAGS_DEFAULT
+#if PY_MAJOR_VERSION >= 3
+    // Py_TPFLAGS_HAVE_ITER is deprecated in Python3.
+#else
+    | Py_TPFLAGS_HAVE_ITER
+#endif
+    , /* tp_flags */
   "FastCSV Writer Object",    /* tp_doc */
   0,                          /* tp_traverse */
   0,                          /* tp_clear */

--- a/tests.py
+++ b/tests.py
@@ -4,7 +4,7 @@ from __future__ import division, absolute_import, print_function, unicode_litera
 import __main__
 import os.path
 import re
-from unittest.loader import _CmpToKey
+import functools
 from setuptools.command.test import ScanningLoader
 
 class RegexpPrefixLoader(ScanningLoader):
@@ -18,9 +18,10 @@ class RegexpPrefixLoader(ScanningLoader):
             if not pattern: pattern = self.testMethodPattern
             return re.match(pattern, attrname) and \
                 hasattr(getattr(testCaseClass, attrname), '__call__')
-        testFnNames = filter(isTestMethod, dir(testCaseClass))
+        testFnNames = list(filter(isTestMethod, dir(testCaseClass)))
         if self.sortTestMethodsUsing:
-            testFnNames.sort(key=_CmpToKey(self.sortTestMethodsUsing))
+            testFnNames.sort(
+                key=functools.cmp_to_key(self.sortTestMethodsUsing))
         return testFnNames
 
 def load_tests(loader, tests, pattern):

--- a/writer_test.py
+++ b/writer_test.py
@@ -57,6 +57,8 @@ class WriterTest(unittest.TestCase):
 
     def it_converts_non_unicode_to_unicode(self):
         class A(object):
+            def __str__(self):
+                return "A object"
             def __unicode__(self):
                 return "A object"
 
@@ -85,6 +87,8 @@ class WriterTest(unittest.TestCase):
 
     def it_raises_ValueError_if_a_cell_value_cannot_be_converted_to_unicode(self):
         class A(object):
+            def __str__(self):
+                raise Exception
             def __unicode__(self):
                 raise Exception
 

